lua/lspkind/init.lua

```diff
@@ -1,185 +1,309 @@
+      Number = '',
+      Boolean = '蘒',
+      Array = '',
+      Object = '',
+      Key = '',
+      Null = 'ﳠ',
+    },
   },
 }
 
-local kind_order = {
-  'Text', 'Method', 'Function', 'Constructor', 'Field', 'Variable', 'Class', 'Interface', 'Module',
-  'Property', 'Unit', 'Value', 'Enum', 'Keyword', 'Snippet', 'Color', 'File', 'Reference', 'Folder',
-  'EnumMember', 'Constant', 'Struct', 'Event', 'Operator', 'TypeParameter'
-}
-local kind_len = 25
+-- Expose presets in public API
+M.presets = kinds.presets
 
-local function get_symbol(kind) 
-    local symbol = lspkind.symbol_map[kind]
-    return symbol or ''
+-- Get symbol info
+function H.get_symbol(list, order, item)
+  local kind = {}
+  for _kind, _idx in pairs(order) do
+    if type(item) == 'string' and item == _kind then
+      kind.name = _kind
+      kind.icon = list[_idx]
+      kind.idx = _idx
+    elseif type(item) == 'number' and item == _idx then
+      kind.name = _kind
+      kind.icon = list[item]
+      kind.idx = item
+    end
+  end
+  return kind
 end
 
-local modes = { 
-    ['text'] = function(kind)
-        return kind
-    end, 
-    ['text_symbol'] = function(kind)
-        local symbol = get_symbol(kind)
-        return fmt("%s %s", kind, symbol)
-    end, 
-    ['symbol_text'] = function(kind)
-        local symbol = get_symbol(kind)
-        return fmt("%s %s", symbol, kind)
-    end, 
-    ['symbol']  = function(kind)
-        local symbol = get_symbol(kind)
-        return fmt("%s", symbol)
-    end 
+-- Symbol modes
+H.modes = {
+  text = function(kind)
+    return kind.name
+  end,
+  text_symbol = function(kind)
+    return fmt('%s %s', kind.name, kind.icon)
+  end,
+  symbol = function(kind)
+    return fmt('%s', kind.icon)
+  end,
+  symbol_text = function(kind)
+    return fmt('%s %s', kind.icon, kind.name)
+  end,
 }
 
--- default true
--- deprecated
-local function opt_with_text(opts)
-  return opts == nil or opts['with_text'] == nil or opts['with_text']
-end
+-- Default options
+H.defaults = {
+  mode = 'symbol_text',
+  symbols = 'default',
+}
 
--- default 'symbol'
-local function opt_mode(opts) 
-    local mode = 'symbol'
-    if opt_with_text(opts) and opts ~= nil and opts['mode'] == nil then
-        mode = 'symbol_text'
-    elseif opts ~= nil and opts['mode'] ~= nil then
-        mode = opts['mode']
-    end
-    return mode
+-- Format symbols
+function H.fmt_kind(list, order, mode, _kind, icon)
+  local kitem = {
+    name = H.get_symbol(list, order, _kind).name,
+    icon = icon,
+  }
+  return H.modes[mode](kitem)
 end
 
--- default 'default'
-local function opt_preset(opts)
-  local preset
-  if opts == nil or opts['preset'] == nil then
-    preset = 'default'
-  else
-    preset = opts['preset']
+-- Define symbols
+function H.set_symbols(ktype, order, symbols, mode)
+  for kind, icon in pairs(symbols) do
+    if ktype[kind] then
+      local idx = ktype[kind]
+      ktype[idx] = H.fmt_kind(ktype, order, mode, kind, icon)
+    end
```

Comment:
`H.set_symbols` is called from L257-258. Actually in this helper function you don't need `symbols` and `mode` as parameter. Why don't you directly pull from `M` or `opts`?

Reply:
`symbols` is required as it is obtained by merging what the user defined or default
`mode` is required on return of the formatted symbol, also coming from default or user definition

Reply:
> Wait, why do you need to change those table?

If I don't change them, how would I pass the symbols to nvim to recognize them?

Reply:
> If that's the case why are you hard-coding it? That's even more strange. I still believe it is not necessary -- please see comments around L238.

I think it's better to have the positions defined by the code, because in case of accidental alteration or removal of the position data in the table of vim.lsp.protocol the plugin will still be able to define the symbols